### ISSUE TITLE
Timetag server side support

### ISSUE THREAD

**Issue by rseymour** [johnelse/ocaml-osc#10]
I was wondering how much work timetag support would be for this library.  ocaml-lo doesn't support timetags either, and I'd really like to have an ocaml based server with timetags working for some time based math I want to try out.

Any roadmap for this, could it be easy enough for a relative newbie like me to work out?

**Comment by johnelse** [comment]
Hey, I've pushed some changes to add timetag support to https://github.com/johnelse/ocaml-osc/tree/timetag

If you're using opam, you can install it with

```
opam pin add osc git://github.com/johnelse/ocaml-osc#timetag
```

It actually wasn't much work, as the code for encoding/decoding timetags was already there as part of the bundle support.

I've only done minimal testing so far, so let me know if it works for you. If it does I'll write some tests and make a proper release.

**Comment by rseymour** [comment]
I finally got a chance to test this, but I'm not seeing anything coming through.  I grabbed the code from the pin.  Using TouchOSC and oscdump I get

> 
> oscdump osc.udp://:7777   
> dd9d15a7.922f48c2 /1/fader3 f 0.244478   
> dd9d15a7.9235b4ed /1/fader3 f 0.288989   
> dd9d15a7.9237585b /1/fader3 f 0.326378   
> dd9d15a7.923e5b84 /1/fader3 f 0.351305   
> dd9d15a7.93a37691 /1/fader3 f 0.374451   

Now with the code I wrote in ocaml I don't get any timestamps still ... hard to say exactly what it is missing:

> /1/fader1
> Float32 0.098481
> /1/fader1
> Float32 0.098481

Here's the code I was using to produce that output, could be user error, but it typechecks and compiles...:
https://gist.github.com/rseymour/1d17c774cff9764265cc3345f1f62688

**Comment by johnelse** [comment]
Ah apologies, I think I've misunderstood - what I've done is added support for the timetag argument type as described here: http://opensoundcontrol.org/spec-1_0.

This is a type additional to the standard int32, float32, string and blob types, but it still requires the client to add a timetag argument to the message.

What I could do is add a function to convert a unix time into a timetag, so you could do something like

```
Osc.Types.make_timetag (Unix.gettimeofday ())
```

at the start of your server callback.

This will only give you the server time though - the only way to get the time direct from the client is to have it add a timetag argument to each message.

Note that bundles include timetags as part of the OSC 1.0 spec, however this is not intended to represent the current time - it's the time at which the OSC methods corresponding to the including messages should be called, i.e. potentially a time in the future.

**Comment by johnelse** [comment]
I hope this helps - let me know your use case and I'll see what I can do :-)

**Comment by rseymour** [comment]
It may be that Immediate is the timetag I want, just looking at the oscdump source code that may be the case.  I'll see if I can get a chance to try that.  What I'm looking to do with all this is calculate the derivative of position, and I'd like to get as close as possible to the timestamps on the device instead of whenever they get across the network.  Barring that I think the "make_timetag" option is my best choice.  I'll keep exploring.  Thanks a ton for your library and your assistance, I definitely owe you a beverage of your choice.

**Comment by johnelse** [comment]
Hey no problem, glad you're finding it useful!

Immediate is actually just a special timetag with value 0 seconds and 1 milliseconds. When a server receives a bundle with this timetag, it's supposed to call the corresponding method(s) ASAP. The "Temporal Semantics" section of the document linked above gives some more details.

If you're in control of the client side, it would be possible to write a client that adds a timetag representing its current time to all messages it sends.

**Comment by rseymour** [comment]
Sorry I let this hang for a while.  It turns out none of the two iOS apps I tried send the timetag.  I'm pretty confident this code works, so the issue can be resolved.  Thanks again for your quick resolution.  I will let you know if I find an app which natively sends timetags.

**Comment by johnelse** [comment]
Sounds good, thanks for the update. FYI I have merged this branch into a new opam release, 0.1.4, so you shouldn't have to use the pin any more.
